pkg/filters/proxy/loadbalance.go

```diff
@@ -194,11 +202,61 @@ func newHeaderHashLoadBalancer(servers []*Server, key string) *headerHashLoadBal
 }
 
 // ChooseServer implements the LoadBalancer interface.
-func (lb *headerHashLoadBalancer) ChooseServer(req *httpprot.Request) *Server {
+func (lb *headerHashLoadBalancer) ChooseServer(spCtx *serverPoolContext) *Server {
 	if len(lb.Servers) == 0 {
 		return nil
 	}
-	v := req.HTTPHeader().Get(lb.key)
+	v := spCtx.req.HTTPHeader().Get(lb.key)
+	hash := fnv.New32()
+	hash.Write([]byte(v))
+	return lb.Servers[hash.Sum32()%uint32(len(lb.Servers))]
+}
+
+// stickySessionsLoadBalancer does load balancing based on sticky sessions.
+type stickySessionsLoadBalancer struct {
+	BaseLoadBalancer
+	SessionCookieName    string
+	SessionCookieExpires time.Duration
+}
+
+func newStickySessionsLoadBalancer(servers []*Server, cookieName string, cookieExpires string) *stickySessionsLoadBalancer {
+	lb := &stickySessionsLoadBalancer{
+		BaseLoadBalancer: BaseLoadBalancer{
+			Servers: servers,
+		},
+	}
+	if cookieName == "" {
+		cookieName = "EASEGRESS_SESSIONID"
+	}
+	lb.SessionCookieName = cookieName
+	if cookieExpires != "" {
+		if d, err := time.ParseDuration(cookieExpires); err == nil {
+			lb.SessionCookieExpires = d
+		}
+	}
+	return lb
+}
+
+func (lb *stickySessionsLoadBalancer) ChooseServer(spCtx *serverPoolContext) *Server {
```

Comment:
This could fail in the case of service discovery, because the order of the available servers could change from time to time.
I think the cookie value should be something related to the url of the servers.